### PR TITLE
Add/Refactor PushProtocol Service

### PR DESCRIPTION
**Pull request by AndiAkasha**
## Description
New service implemented called PushProtocol which contains several methods that are required in order to initialise, change settings, get notifications.
## Checklist

- [ ] I have read the **README** document
- [ ] I have read the **CONTRIBUTING** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] My commit message follows [Conventional Commit Guideline](https://conventionalcommits.org/)

### CODE REVIEW

File: libs/typings/src/sdk/notification.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  .object({
+    readonly: z.boolean(),
+  })
+  .default({
+    readonly: true,
+  });
+
+export type InitializePushProtocolOptions = z.infer<typeof InitializeOptionsSchema>;
+
+const NotificationSettingTypeSchema = z.object({
```
Comment (kenshyx) [inline libs/typings/src/sdk/notification.ts]: typings are ok but we should avoid keeping also the validators here(extra overhead)

Comment (kenshyx) [reply libs/typings/src/sdk/notification.ts]: I mean in the `typings` package

File: libs/sdk/src/common/env.ts [hunk 2]
```diff
@@ -14,6 +14,7 @@ export const env = createEnv({
         message: 'DID must start with did:',
       }),
     PUBLIC_WALLET_CONNECT_PROJECT_ID: z.string().min(12),
+    PUBLIC_PUSH_PROTOCOL_CHANNEL_ID: z.string().min(42),
```
Comment (kenshyx) [inline libs/sdk/src/common/env.ts:17]: update also the .ev.example file with this

File: libs/sdk/src/common/notification.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+      // If the settings are sent without order or there are some opt-in missing PushProtocol sets all the opt-in to false
+      throw new Error(
+        'Settings must contain all the opt-ins available. Please be aware that the order of the opt-in sent matter',
+      );
+    await this.notificationsWriteClient.notification.subscribe(this._notificationChannelId, {
+      settings: newSettings,
+    });
+  }
+
+  async getNotifications(page: number = 1, limit: number = 50) {
```
Comment (kenshyx) [inline libs/sdk/src/common/notification.ts]: add validators also here

File: libs/sdk/src/common/notification/notification.ts [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+    } else {
+      /**
+       * We initialize with a signer in order to send notification or change the opt-in/settings
+       * If the user clicks cancel, this method will throw an Error
+       */
+      await this.initializeWritableClient();
+    }
+  }
+
+  async listenToNotificationEvents() {
```
Comment (kenshyx) [inline libs/sdk/src/common/notification/notification.ts:101]: add some docs about this method

File: libs/sdk/src/common/notification/notification.ts [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+
+  async stopListeningToNotificationEvents() {
+    if (this._notificationsStream) {
+      await this._notificationsStream.disconnect();
+      this._notificationsStream.removeAllListeners(CONSTANTS.STREAM.NOTIF);
+      this._notificationsStream = undefined;
+    }
+  }
+
+  async getSettingsOfChannel(): Promise<ChannelSettings[]> {
```
Comment (kenshyx) [inline libs/sdk/src/common/notification/notification.ts:154]: docs

File: libs/sdk/src/common/notification/notification.ts [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+    const response = await this.notificationsClient.channel.info(this._notificationChannelId);
+    const parseResponse = ChannelInfoResponseSchema.safeParse(response);
+    if (!parseResponse.success) throw new Error('User Settings unable to parse');
+
+    return parseResponse.data;
+  }
+
+  async getSettingsOfUser(): Promise<UserSettingType[]> {
```
Comment (kenshyx) [inline libs/sdk/src/common/notification/notification.ts:173]: docs

File: libs/sdk/src/common/notification/notification.ts [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+    } else if (Notification.permission !== 'denied') {
+      // We need to ask the user for permission
+      const perm = await Notification.requestPermission();
+      return perm === 'granted';
+    }
+
+    return false;
+  }
+
+  private parseNotificationData(
```
Comment (kenshyx) [inline libs/sdk/src/common/notification/notification.ts:350]: docs

File: libs/sdk/src/common/notification/notification.ts [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+    }
+
+    // Return the parsed index and data
+    return {
+      channelIndex: indexOfOption,
+      data,
+    };
+  }
+
+  private async initializeWritableClient(): Promise<void> {
```
Comment (kenshyx) [inline libs/sdk/src/common/notification/notification.ts:419]: docs

File: libs/sdk/src/common/notification/notification.ts [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+    }
+    notification.timestamp = new Date(notification.epoch);
+    notification.isUnread = latestStoredNotificationID
+      ? notification.payload_id > latestStoredNotificationID
+      : true;
+
+    return notification;
+  }
+
+  private parseMetaData(metaData: AdditionalMetadata): NotificationParsedMetaData {
```
Comment (kenshyx) [inline libs/sdk/src/common/notification/notification.ts:381]: docs

**Review by kenshyx — APPROVED** [review]

**Review by kunstefix — APPROVED** [review]

**Pull request merged**